File: java-suite/src/main/java/de/fraunhofer/iem/generics/upperBounds/inGenericFunctions/MultipleUpperBound.java
package de.fraunhofer.iem.generics.upperBounds.inGenericFunctions;

import de.fraunhofer.iem.generics.wildCardTypeArgument.DifferentTypedPair;

public class MultipleUpperBound {
    /**
     * Demonstrates the multiple upper bound in the generic function in Java
     * <p>
     * Note:
     * 1. If multiple upper bounds are given, then that means a Class should extends all of those upper bound then only
     * its valid otherwise compilation error
     *
     * @author Ranjith Krishnamurthy
     */
    public static <A extends Number, B extends Comparable<B> & CharSequence> DifferentTypedPair<A, B> createDifferentTypedPair(A first, B second) {
        return new DifferentTypedPair<>(first, second);
    }

    /**
     * Demonstrates the usage of the multiple upper bounded generic function in Java
     *
     * @author Ranjith Krishnamurthy
     */
    public static void main(String[] args) {
        System.out.println(createDifferentTypedPair(254564, "Ranjith"));
    }
}

File: kotlin-suite/src/main/kotlin/de/fraunhofer/iem/generics/upperBounds/inGenericFunctions/MultipleUpperBound.kt
package de.fraunhofer.iem.generics.upperBounds.inGenericFunctions

import de.fraunhofer.iem.uniqueToKotlin.declarationSiteVarianceAndTypeProjetion.typeProjection.DifferentTypedPair

/**
 * Demonstrates the multiple upper bound in the generic function in Kotlin
 *
 * @author Ranjith Krishnamurthy
 */
fun <A : Number, B> createDifferentTypedPair1(first: A, second: B): DifferentTypedPair<A, B>
        where B : Comparable<B>,
              B : CharSequence {
    return DifferentTypedPair<A, B>(first, second);
}

/**
 * Demonstrates the usage of the multiple upper bounded generic function in Kotlin
 *
 * @author Ranjith Krishnamurthy
 */
fun main() {
    println(createDifferentTypedPair1<Int, String>(2646, "Ranjith"))
}